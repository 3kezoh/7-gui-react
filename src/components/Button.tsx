import { cn } from "../utils";

type ButtonProps = React.ComponentPropsWithoutRef<"button">;

export function Button(props: ButtonProps) {
  const { className, disabled, ...rest } = props;

  const buttonClassName = cn(
    "border border-black px-2",
    className,
    {
      "transition-colors hover:text-white hover:bg-black": !disabled,
    },
    {
      "cursor-not-allowed bg-gray-200 text-gray-400 border-gray-400": disabled,
    },
  );

  return (
    <button
      className={buttonClassName}
      disabled={disabled}
      type="button"
      {...rest}
    />
  );
}
