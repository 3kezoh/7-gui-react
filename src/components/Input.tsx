import { cn } from "../utils";

type InputProps = React.ComponentPropsWithoutRef<"input">;

export function Input(props: InputProps) {
  const { className, name, ...rest } = props;
  const inputClassName = cn("border border-black w-full max-w-xs", className);

  return (
    <input
      className={inputClassName}
      type="text"
      id={name}
      name={name}
      {...rest}
    />
  );
}
