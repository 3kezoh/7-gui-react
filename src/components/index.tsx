export * from "./Counter";
export * from "./FlightBooker";
export * from "./TemperatureConverter";
