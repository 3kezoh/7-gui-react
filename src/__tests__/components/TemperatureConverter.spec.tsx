import { render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { TemperatureConverter } from "../../components";

describe("TemperatureConverter", () => {
  it("should render the temperature converter", () => {
    render(<TemperatureConverter />);

    expect(screen.getByTestId("temperatureConverter")).toBeInTheDocument();
  });

  it("should render 100 Celsius 212 Fahrenheit", () => {
    render(<TemperatureConverter celsius={100} />);

    const celsiusInput = screen.getByTestId("celsius");
    const fahrenheitInput = screen.getByTestId("fahrenheit");

    expect(celsiusInput).toHaveValue(100);
    expect(fahrenheitInput).toHaveValue(212);
  });

  it("should render 23 Fahrenheit and -5 Celsius", () => {
    render(<TemperatureConverter fahrenheit={23} />);

    const celsiusInput = screen.getByTestId("celsius");
    const fahrenheitInput = screen.getByTestId("fahrenheit");

    expect(celsiusInput).toHaveValue(-5);
    expect(fahrenheitInput).toHaveValue(23);
  });

  it("should convert -13e3 Celsius to -23368 Fahrenheit", async () => {
    const user = userEvent.setup();

    render(<TemperatureConverter />);

    const celsiusInput = screen.getByTestId("celsius");
    const fahrenheitInput = screen.getByTestId("fahrenheit");

    await user.clear(celsiusInput);

    await user.type(celsiusInput, "-13e3");

    expect(fahrenheitInput).toHaveValue(-23368);
  });

  it("should convert -13 Fahrenheit to -25 Celsius", async () => {
    const user = userEvent.setup();

    render(<TemperatureConverter />);

    const celsiusInput = screen.getByTestId("celsius");
    const fahrenheitInput = screen.getByTestId("fahrenheit");

    await user.clear(fahrenheitInput);

    await user.type(fahrenheitInput, "-13");

    expect(celsiusInput).toHaveValue(-25);
  });

  it("should not convert a non-numeric value to Fahrenheit", async () => {
    const user = userEvent.setup();

    render(<TemperatureConverter />);

    const celsiusInput = screen.getByTestId("celsius");
    const fahrenheitInput = screen.getByTestId("fahrenheit");

    await user.clear(celsiusInput);

    await user.type(celsiusInput, "abc");

    expect(fahrenheitInput).toHaveValue(32);
  });

  it("should not convert a non-numeric value to Celsius", async () => {
    const user = userEvent.setup();

    render(<TemperatureConverter />);

    const celsiusInput = screen.getByTestId("celsius");
    const fahrenheitInput = screen.getByTestId("fahrenheit");

    await user.clear(fahrenheitInput);

    await user.type(fahrenheitInput, "abc");

    expect(celsiusInput).toHaveValue(0);
  });

  it("should not convert an empty value to Fahrenheit", async () => {
    const user = userEvent.setup();

    render(<TemperatureConverter />);

    const celsiusInput = screen.getByTestId("celsius");
    const fahrenheitInput = screen.getByTestId("fahrenheit");

    await user.clear(celsiusInput);

    expect(fahrenheitInput).toHaveValue(32);
  });

  it("should not convert an empty value to Celsius", async () => {
    const user = userEvent.setup();

    render(<TemperatureConverter />);

    const celsiusInput = screen.getByTestId("celsius");

    const fahrenheitInput = screen.getByTestId("fahrenheit");

    await user.clear(fahrenheitInput);

    expect(celsiusInput).toHaveValue(0);
  });

  it("should match the snapshot", () => {
    const { asFragment } = render(<TemperatureConverter />);

    expect(asFragment()).toMatchSnapshot();
  });
});
