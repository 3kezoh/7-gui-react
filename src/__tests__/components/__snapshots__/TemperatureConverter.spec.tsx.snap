// Vitest Snapshot v1

exports[`TemperatureConverter > should match the snapshot 1`] = `
<DocumentFragment>
  <div
    class="temperatureConverter"
    data-testid="temperatureConverter"
  >
    <input
      class="temperature"
      data-testid="celsius"
      name="celsius"
      type="number"
      value="0"
    />
    <span>
      Celsius =
    </span>
    <input
      class="temperature"
      data-testid="fahrenheit"
      name="fahrenheit"
      type="number"
      value="32"
    />
    <span>
      Fahrenheit
    </span>
  </div>
</DocumentFragment>
`;
